### DISCUSSION TITLE
Add ControlNet image file name to prompt when batch processing?

### DISCUSSION THREAD

**Discussion by FugueSegue** [vladmandic/sdnext#1046] [Ideas]
Is there any way to add a checkbox to the ControlNet batch tab so that it adds the name of each file to the beginning of the prompt?

I am trying to generate a vast number of images that will subsequently be used as a dataset for a style. For example, I have three pictures that I want to batch process in ControlNet. One is a photo of a car, another of a truck, another of a bus. I want to render all of them with prompts that include, "[art by van gogh|art by greg rutkowski|art by alphonse mucha]" in order to combine those artist's styles. But I also need to include "car" or "truck" or "bus" in the prompts that use those images when they are each used by ControlNet.

I tried to use Prompts From File script but it doesn't work the way I expected. The text I fed to the script looked like this:

car, [art by van gogh|art by greg rutkowski|art by alphonse mucha]
truck, [art by van gogh|art by greg rutkowski|art by alphonse mucha]
bus, [art by van gogh|art by greg rutkowski|art by alphonse mucha]

The result is that it uses the car prompt three times. Once using the car image in ControlNet, then using the truck image, then using the bus image. It repeats this process for the truck prompt, and then again for the bus prompt. I only want to generate three images but it instead renders nine. Three of those nine images are exactly what I want but this is a huge waste of rendering time. Especially if I'm trying to process more than a hundred images.

If I could simply have ControlNet read the file name of each image in the batch and add it to the beginning of the prompt, I believe it would solve this problem. I suppose the question is whether or not such a feature would have use beyond the narrow scope of what I want to do.

Or is there another way to do what I'm trying to do?

**Comment by Tillerz** [comment]
Try the extension sd_dynamic prompts. There's built in stuff like the x y z prompt stuff you can use too, but this is a really helpful extension, as it allows RANDOM inserts, too.

You define your stuff in wildcard files, which work recursively, too. So you can use wildcards from other wildcard files:

extensions\sd-dynamic-prompts\wildcards\vehicle.txt
```
__color__ __type__
```

extensions\sd-dynamic-prompts\wildcards\color.txt
```
red
blue
green
```

extensions\sd-dynamic-prompts\wildcards\type.txt
```
car
truck
bus
```

Now you can add `__vehicle__` to the prompt and enable the extension to get stuff like red bus or green truck added to your prompt:

![dynamic prompt](https://i.gyazo.com/e0ea4adccaa57b2bbb0d6f697cea24da.png)

If you have "Combinatorial generation" **enabled**, it will generate images with ALL combinations by iterating over them.

If you have it **disabled** as in the screenshot, it will just RANDOMLY pick one of each for how many images you generate. Even if you just generate ONE image, it will pick randomly, not just always the first one.


So you would just use

```
__type__, [art by van gogh|art by greg rutkowski|art by alphonse mucha]
```

for your usecase and enable **Combinatorial generation** to get an image generated for each vehicle type.

**Reply by FugueSegue** [reply]
Thank you for the excellent reply. I think I had seen Dynamic Prompts mentioned elsewhere but I had forgotten about its existence. I had never used it before. Based on your recommendation, I've been experimenting with it for the last few hours.

Unfortunately, I can't figure out how to use it to do the specific task I want to do. I'll try to explain again using my example of the car, truck, and bus.

I want to generate three images. I'm using the ControlNet reference-only model and using the batch option. The specified Input Directory for the batch contains three photos: one of a car, one of a truck, and one of a bus. The files are named, 01_car.png, 02_truck.png, and 03_bus.png. I can specify "[art by van gogh|art by greg rutkowski|art by alphonse mucha]" in the prompt but the reference-only model requires something in the prompt that describes each respective image. Otherwise it makes some sort of random guess based on the limited range of what each artist style can generate. When ControlNet processes the image of a car, I need to explicitly state in the prompt that it is a "car".

What I need to have it do is generate three images. Each image should be generated with these three prompts and ControlNet conditions:

Image 1
Prompt: **car**, [art by van gogh|art by greg rutkowski|art by alphonse mucha]
ControlNet: reference-only using an image of a **car**.

Image 2
Prompt: **truck**, [art by van gogh|art by greg rutkowski|art by alphonse mucha]
ControlNet: reference-only using an image of a **truck**.

Image 3
Prompt: **bus**, [art by van gogh|art by greg rutkowski|art by alphonse mucha]
ControlNet: reference-only using an image of a **bus**.

I can't seem to figure out how to do this with Dynamic Prompts. I'm even trying to use Jinja2 templates. But I can't figure out how to iterate in the manner I want. The trouble I'm having is that a Dynamic Prompt runs through any manner of variations that I specify for each ControlNet batch image. It goes through all the permutations and then repeats the process for each image. That's not what I want. My intuition tells me that Dynamic Prompts cannot do what I'm trying to do.

I hope this helps clarify my goal. Thank you again for your help.

**Comment by FugueSegue** [comment]
I just realized that I should not have suggested an idea for ControlNet here. I should post my idea over in the discussion forum for that specific extension.
